### PR TITLE
Feat 2857:  Implement Gen2 Support for `ibm_database` datasource

### PR DESCRIPTION
**Pull request by pushpalathaeg**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Ticket : https://github.ibm.com/compose/App/issues/2857

## Overview

This documentation covers the enhancement to the IBM Cloud Terraform Provider that adds support for **Generation 2 (Gen2) database instances** in the `ibm_database` data source. This update enables users to seamlessly query both Classic and Gen2 database deployments using the same data source with automatic detection and appropriate handling of generation-specific features.

---
### Gen2 Database Support in Data Source

The `ibm_database` data source now **automatically detects** whether a database instance is Classic or Gen2 and retrieves the appropriate information. This provides a unified interface for querying database instances regardless of their generation.

### Key Features

✅ **Automatic Detection**: The provider automatically identifies whether a database is Classic or Gen2 based on the plan type  
✅ **Unified Interface**: Use the same `ibm_database` data source for both Classic and Gen2 instances  
✅ **Backward Compatible**: Existing Terraform configurations continue to work without modification  
✅ **Proper State Management**: Gen2-specific behavior ensures clean state management when switching between instance types  
✅ **Comprehensive Testing**: Full test coverage for Gen2 scenarios including edge cases  

---

### Classic Databases (Standard Plans)

- **Deployment Model**: Traditional IBM Cloud Databases architecture
- **Plan Names**: `standard`, `enterprise`, `platinum`
- **Features**: Full feature set including:
  - Default admin user management
  - IP allowlists for network access control
  - Auto-scaling configuration
  - Direct user management through API
  - Configuration schema access

### Gen2 Databases (Standard-Gen2 Plans)

- **Deployment Model**: Next-generation architecture with enhanced capabilities
- **Plan Names**: `standard-gen2`, `enterprise-gen2`
- **Benefits**:
  - Enhanced performance and scalability
  - Improved resource isolation
  - Modern infrastructure
  - Better integration with IBM Cloud platform services
- **Management Model**: Different approach to credentials and access control

---

### Backend Selection Pattern

The data source uses a **strategy pattern** to handle different database generations:

```
User Query → findInstance() → Detect Plan Type → Select Backend
                                                      ↓
                                    ┌─────────────────┴─────────────────┐
                                    ↓                                   ↓
                          Classic Backend                        Gen2 Backend
                          (Full API Access)                (Metadata-based Access)
```

### Implementation Components

1. **`data_source_ibm_database.go`**: Main data source with backend selection logic
2. **`data_source_ibm_database_classic.go`**: Classic database backend implementation
3. **`data_source_ibm_database_gen2.go`**: Gen2 database backend implementation
4. **`helpers.go`**: Shared utility functions for both backends
5. **Test Files**: Comprehensive test coverage for both generations

---

### Attributes Not Available in Gen2

The following attributes are **not supported** for Gen2 database instances and will return empty/zero values:

| Attribute | Classic Behavior | Gen2 Behavior | Reason |
|-----------|------------------|---------------|--------|
| `adminuser` | Returns default admin username | Returns empty string | No default admin user in Gen2 |
| `adminpassword` | Returns admin password | Returns empty string | Credentials managed via service keys |
| `users` | Returns list of database users | Returns empty list `[]` | User management through service keys |
| `allowlist` | Returns IP allowlist entries | Returns empty list `[]` | Network access via VPC/CBR |
| `auto_scaling` | Returns auto-scaling config | Returns empty list `[]` | Not currently supported |
| `configuration_schema` | Returns JSON schema | Returns empty string | Not available in Gen2 |
| `backup_encryption_key_crn` | Returns backup encryption key | Not set | Different encryption model |

### Attributes Available in Gen2

These attributes work identically in both Classic and Gen2:

✅ `name` - Database instance name  
✅ `service` - Database service type (e.g., `databases-for-postgresql`)  
✅ `plan` - Service plan (e.g., `standard-gen2`)  
✅ `status` - Instance status  
✅ `version` - Database version  
✅ `location` - Region/location  
✅ `resource_group_id` - Resource group identifier  
✅ `guid` - Unique instance identifier  
✅ `groups` - Scaling group configurations (memory, disk, CPU)  
✅ `platform_options.disk_encryption_key_crn` - Disk encryption key  
✅ `tags` - Instance tags  

---

### Example 1: Basic Data Source Query

Works for both Classic and Gen2 databases:

```hcl
# Query any database instance
data "ibm_database" "database" {
  name = "my-database-instance"
}

# Access common attributes
output "database_info" {
  value = {
    version  = data.ibm_database.database.version
    location = data.ibm_database.database.location
    plan     = data.ibm_database.database.plan
    status   = data.ibm_database.database.status
  }
}
```

### Example 2: Gen2 Database with Credential Management

For Gen2 databases, use `ibm_resource_key` for credentials:

```hcl
# Query Gen2 database instance
data "ibm_database" "my_gen2_db" {
  name = "my-postgres-gen2"
}

# Create service credentials for Gen2 database
resource "ibm_resource_key" "db_credentials" {
  name                 = "my-db-credentials"
  resource_instance_id = data.ibm_database.my_gen2_db.id
  role                 = "Administrator"
}

# Access connection details
output "database_connection" {
  value = {
    host     = ibm_resource_key.db_credentials.credentials["connection.postgres.hosts.0.hostname"]
    port     = ibm_resource_key.db_credentials.credentials["connection.postgres.hosts.0.port"]
    database = ibm_resource_key.db_credentials.credentials["connection.postgres.database"]
    username = ibm_resource_key.db_credentials.credentials["connection.postgres.authentication.username"]
    password = ibm_resource_key.db_credentials.credentials["connection.postgres.authentication.password"]
  }
  sensitive = true
}
```

### Example 3: Querying with Resource Group Filter

```hcl
data "ibm_resource_group" "group" {
  name = "production"
}

data "ibm_database" "database" {
  name              = "my-database"
  resource_group_id = data.ibm_resource_group.group.id
}
```

### Example 4: Querying with Location Filter

```hcl
data "ibm_database" "database" {
  name     = "my-database"
  location = "ca-mon"
}

# Useful when multiple databases share the same name in different regions
```

### Example 5: Accessing Scaling Groups Information

```hcl
data "ibm_database" "database" {
  name = "my-database"
}

# Memory allocation
output "memory_allocation_mb" {
  value = data.ibm_database.database.groups[0].memory[0].allocation_mb
}

# Disk allocation
output "disk_allocation_mb" {
  value = data.ibm_database.database.groups[0].disk[0].allocation_mb
}

# CPU allocation (if available)
output "cpu_allocation_count" {
  value = data.ibm_database.database.groups[0].cpu[0].allocation_count
}

# Check if resources are adjustable
output "can_scale_memory" {
  value = data.ibm_database.database.groups[0].memory[0].is_adjustable
}
```

### Example 6: Conditional Logic Based on Plan Type

```hcl
data "ibm_database" "database" {
  name = "my-database"
}

locals {
  is_gen2 = can(regex("gen2", data.ibm_database.database.plan))
}

# Use different credential management based on generation
resource "ibm_resource_key" "credentials" {
  count = local.is_gen2 ? 1 : 0
  
  name                 = "gen2-credentials"
  resource_instance_id = data.ibm_database.database.id
  role                 = "Administrator"
}

output "credential_source" {
  value = local.is_gen2 ? "Using ibm_resource_key for Gen2" : "Using adminuser for Classic"
}
```

### Example 7: Platform Options and Encryption

```hcl
data "ibm_database" "database" {
  name = "my-encrypted-database"
}

# Access disk encryption key
output "disk_encryption_key" {
  value = try(
    data.ibm_database.database.platform_options[0].disk_encryption_key_crn,
    "No encryption key configured"
  )
}
```

### Input Parameters

| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `name` | String | Yes | Database instance name |
| `resource_group_id` | String | No | Filter by resource group ID |
| `location` | String | No | Filter by region/location |
| `service` | String | No | Filter by service type |

### Output Attributes - Common (Both Generations)

| Attribute | Type | Description |
|-----------|------|-------------|
| `id` | String | Resource instance ID |
| `name` | String | Database instance name |
| `service` | String | Database service type |
| `plan` | String | Service plan (includes `-gen2` suffix for Gen2) |
| `status` | String | Instance status (active, provisioning, etc.) |
| `version` | String | Database version |
| `location` | String | Region/location |
| `resource_group_id` | String | Resource group ID |
| `guid` | String | Unique instance identifier |
| `resource_name` | String | Resource name |
| `resource_crn` | String | Cloud Resource Name |
| `resource_status` | String | Resource status |
| `resource_group_name` | String | Resource group name |
| `resource_controller_url` | String | Dashboard URL |
| `tags` | Set(String) | Instance tags |

### Output Attributes - Groups (Both Generations)

| Attribute | Type | Description |
|-----------|------|-------------|
| `groups` | List | Scaling group configurations |
| `groups[].group_id` | String | Group identifier (e.g., "member") |
| `groups[].count` | Integer | Number of members in group |
| `groups[].memory` | List | Memory configuration |
| `groups[].memory[].allocation_mb` | Integer | Current memory allocation in MB |
| `groups[].memory[].minimum_mb` | Integer | Minimum allowed memory |
| `groups[].memory[].step_size_mb` | Integer | Memory scaling increment |
| `groups[].memory[].is_adjustable` | Boolean | Whether memory can be adjusted |
| `groups[].memory[].can_scale_down` | Boolean | Whether memory can scale down |
| `groups[].disk` | List | Disk configuration |
| `groups[].disk[].allocation_mb` | Integer | Current disk allocation in MB |
| `groups[].disk[].minimum_mb` | Integer | Minimum allowed disk |
| `groups[].disk[].step_size_mb` | Integer | Disk scaling increment |
| `groups[].disk[].is_adjustable` | Boolean | Whether disk can be adjusted |
| `groups[].disk[].can_scale_down` | Boolean | Whether disk can scale down |
| `groups[].cpu` | List | CPU configuration |
| `groups[].cpu[].allocation_count` | Integer | Current CPU allocation |
| `groups[].cpu[].minimum_count` | Integer | Minimum allowed CPUs |
| `groups[].cpu[].step_size_count` | Integer | CPU scaling increment |
| `groups[].cpu[].is_adjustable` | Boolean | Whether CPU can be adjusted |
| `groups[].cpu[].can_scale_down` | Boolean | Whether CPU can scale down |

### Output Attributes - Platform Options (Both Generations)

| Attribute | Type | Gen2 Support | Description |
|-----------|------|--------------|-------------|
| `platform_options` | Set | Yes | Platform-specific options |
| `platform_options[].disk_encryption_key_crn` | String | Yes | Disk encryption key CRN |
| `platform_options[].backup_encryption_key_crn` | String | No | Backup encryption key (Classic only) |

### Output Attributes - Classic Only

| Attribute | Type | Gen2 Behavior | Description |
|-----------|------|---------------|-------------|
| `adminuser` | String | Empty string | Default admin username (Classic only) |
| `adminpassword` | String | Empty string | Default admin password (Classic only) |
| `users` | Set | Empty list | Database users (Classic only) |
| `allowlist` | Set | Empty list | IP allowlist (Classic only) |
| `auto_scaling` | List | Empty list | Auto-scaling configuration (Classic only) |
| `configuration_schema` | String | Empty string | Configuration schema JSON (Classic only) |

---

### Test Coverage

The implementation includes comprehensive test coverage for Gen2 scenarios:

#### Test Cases

1. **TestAccIBMDatabaseDataSourceGen2Basic**
   - Validates successful data retrieval for Gen2 instances
   - Verifies common attributes (name, service, plan, location)
   - Confirms Gen2-specific behavior (empty adminuser, users, allowlist)
   - Checks groups information (memory, disk allocations)

2. **TestAccIBMDatabaseDataSourceGen2WithResourceGroupID**
   - Tests filtering by resource group
   - Validates resource group ID is properly set
   - Ensures correct instance is retrieved when multiple exist

3. **TestAccIBMDatabaseDataSourceGen2InvalidInput**
   - Tests error handling for non-existent databases
   - Validates appropriate error messages
   - Ensures graceful failure

4. **TestAccIBMDatabaseDataSourceGen2InvalidID**
   - Tests error handling for malformed identifiers
   - Validates input validation
   - Ensures proper error reporting

</br>
</br>

**`ibm_database` TF Tests for Gen2**

**TF INPUT CONFIG**
```
# Data Source to Read the Gen2 Database
data "ibm_database" "pushpa_tf_gen2-postgres" {
  name              = "my-postgres-gen2-Test"
  location          = "ca-mon"
  service           = "databases-for-postgresql"
}

# Output the Data Source Information
output "database_info" {
  value = {
    # Basic Information
    name    = data.ibm_database.pushpa_tf_gen2-postgres.name
    id      = data.ibm_database.pushpa_tf_gen2-postgres.id
    status  = data.ibm_database.pushpa_tf_gen2-postgres.status
    version = data.ibm_database.pushpa_tf_gen2-postgres.version
    tags    = data.ibm_database.pushpa_tf_gen2-postgres.tags

    # Platform Options (Gen2 specific)
    platform_options = data.ibm_database.pushpa_tf_gen2-postgres.platform_options

    # Configuration
    plan     = data.ibm_database.pushpa_tf_gen2-postgres.plan
    location = data.ibm_database.pushpa_tf_gen2-postgres.location
    service  = data.ibm_database.pushpa_tf_gen2-postgres.service

    # Resource Scaling
    groups = data.ibm_database.pushpa_tf_gen2-postgres.groups
  }
}
```

**TF OUTPUT**
```
ibm-database git:(2857/gen2-datasource-ibm_database-1) ✗ terraform apply
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

database_info = {
  "groups" = tolist([
    {
      "count" = 2
      "cpu" = tolist([
        {
          "allocation_count" = 4
          "can_scale_down" = true
          "is_adjustable" = true
          "minimum_count" = 6
          "step_size_count" = 2
          "units" = "count"
        },
      ])
      "disk" = tolist([
        {
          "allocation_mb" = 10240
          "can_scale_down" = false
          "is_adjustable" = true
          "minimum_mb" = 20480
          "step_size_mb" = 2048
          "units" = "gb"
        },
      ])
      "group_id" = "member"
      "host_flavor" = tolist([
        {
          "hosting_size" = ""
          "id" = "bx3d.4x20"
          "name" = "bx3d.4x20"
        },
      ])
      "memory" = tolist([
        {
          "allocation_mb" = 20480
          "can_scale_down" = true
          "is_adjustable" = true
          "minimum_mb" = 2048
          "step_size_mb" = 256
          "units" = "gb"
        },
      ])
    },
  ])
  "id" = "crn:v1:bluemix:public:databases-for-postgresql:ca-mon:a/40ddc34a953a8c02f10987b59085b60e:7b231067-9ddd-4de9-964a-d7bf84cfdc3f::"
  "location" = "ca-mon"
  "name" = "my-postgres-gen2-Test"
  "plan" = "standard-gen2"
  "platform_options" = toset([
    {
      "backup_encryption_key_crn" = ""
      "disk_encryption_key_crn" = ""
    },
  ])
  "service" = "databases-for-postgresql"
  "status" = "active"
  "tags" = toset([
    "app:myapp",
    "do-not-delete",
    "env:production",
    "gen2-test",
  ])
  "version" = "18"
}
```

</br>
</br>

**`ibm_database` TF Tests for Classic**

**TF INPUT CONFIG**

```
# Data Source to Read the Classic Database
data "ibm_database" "pushpa_tf_classic-postgres" {
  name              = "Databases for PostgreSQL-tn"
  location          = "us-east"
  service           = "databases-for-postgresql"
}

# Output the Data Source Information
output "database_info" {
  value = {
    # Basic Information
    name    = data.ibm_database.pushpa_tf_classic-postgres.name
    id      = data.ibm_database.pushpa_tf_classic-postgres.id
    status  = data.ibm_database.pushpa_tf_classic-postgres.status
    version = data.ibm_database.pushpa_tf_classic-postgres.version
    tags    = data.ibm_database.pushpa_tf_classic-postgres.tags

    # Platform Options (Gen2 specific)
    platform_options = data.ibm_database.pushpa_tf_classic-postgres.platform_options

    # Configuration
    plan     = data.ibm_database.pushpa_tf_classic-postgres.plan
    location = data.ibm_database.pushpa_tf_classic-postgres.location
    service  = data.ibm_database.pushpa_tf_classic-postgres.service

    # Resource Scaling
    groups = data.ibm_database.pushpa_tf_classic-postgres.groups
  }
}
```

**TF OUTPUT**
```
ibm-database git:(2857/gen2-datasource-ibm_database-1) ✗ terraform apply
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

database_info = {
  "groups" = tolist([
    {
      "count" = 2
      "cpu" = tolist([
        {
          "allocation_count" = 4
          "can_scale_down" = true
          "is_adjustable" = true
          "minimum_count" = 0
          "step_size_count" = 2
          "units" = "count"
        },
      ])
      "disk" = tolist([
        {
          "allocation_mb" = 10240
          "can_scale_down" = false
          "is_adjustable" = true
          "minimum_mb" = 10240
          "step_size_mb" = 2048
          "units" = "mb"
        },
      ])
      "group_id" = "member"
      "host_flavor" = tolist([
        {
          "hosting_size" = ""
          "id" = "multitenant"
          "name" = ""
        },
      ])
      "memory" = tolist([
        {
          "allocation_mb" = 8192
          "can_scale_down" = true
          "is_adjustable" = true
          "minimum_mb" = 8192
          "step_size_mb" = 256
          "units" = "mb"
        },
      ])
    },
  ])
  "id" = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:92d35483-3205-4324-adcb-2e2e5b740673::"
  "location" = "us-east"
  "name" = "Databases for PostgreSQL-tn"
  "plan" = "standard"
  "platform_options" = toset([
    {
      "backup_encryption_key_crn" = ""
      "disk_encryption_key_crn" = ""
    },
  ])
  "service" = "databases-for-postgresql"
  "status" = "active"
  "tags" = toset([
    "claasic-test",
    "classic-test-tasks",
    "isl",
  ])
  "version" = "18"
}
```

#### Prerequisites

```bash
# Set required environment variables
export IC_API_KEY="your-ibm-cloud-api-key"

```

Output from acceptance testing:

```
$ go test -v -run "TestAccIBMDatabaseDataSourceGen2Basic" -timeout 120m
=== RUN   TestAccIBMDatabaseDataSourceGen2Basic
=== PAUSE TestAccIBMDatabaseDataSourceGen2Basic
=== CONT  TestAccIBMDatabaseDataSourceGen2Basic
--- PASS: TestAccIBMDatabaseDataSourceGen2Basic (281.23s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database        282.779s


$ go test -v -run "TestAccIBMDatabaseDataSourceGen2InvalidInput" -timeout 120m
=== RUN   TestAccIBMDatabaseDataSourceGen2InvalidInput
--- PASS: TestAccIBMDatabaseDataSourceGen2InvalidInput (4.35s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database        6.390s


$ go test -v -run "TestAccIBMDatabaseDataSourceGen2InvalidInput" -timeout 120m
=== RUN   TestAccIBMDatabaseDataSourceGen2InvalidID
--- PASS: TestAccIBMDatabaseDataSourceGen2InvalidID (2.64s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database        4.577s

```

### CODE REVIEW

**Comment by obai-1** [comment]
This section `State Management During Migration` in the PR description is misleading. I would remove it.

**Comment by pushpalathaeg** [comment]
> This section `State Management During Migration` in the PR description is misleading. I would remove it.

@obai-1  I removed the `State Management During Migration` section

File: website/docs/d/database.html.markdown [hunk 1]
```diff
@@ -75,7 +75,38 @@ In addition to all argument references list, you can access the following attrib
...
+- `adminpassword` - Not available for Gen2 databases. Use `ibm_resource_key` to manage credentials.
+- `users` - User management is not supported. Use `ibm_resource_key` to manage credentials.
+- `allowlist` - IP allowlisting is not supported for Gen2 databases. Use Context-Based Restrictions (`ibm_cbr_rule`) for IP allowlisting in Gen2.
+- `auto_scaling` - Auto-scaling configuration is not currently supported for Gen2 databases.
+- `configuration_schema` - Configuration schema is not available for Gen2 databases.
+- `platform_options.backup_encryption_key_crn` - Backup encryption key is not supported for Gen2 databases (only `disk_encryption_key_crn` is supported).
+
+### Gen2 Example
+```terraform
+data "ibm_database" "postgres_gen2" {
```
Comment (uibm) [inline website/docs/d/database.html.markdown:100]: The documentation example for Gen2 usage here shows  "ibm_database". Is this consistent with the resource documentation fix ? as per #6854

Comment (pushpalathaeg) [reply website/docs/d/database.html.markdown:100]: @uibm  : For Gen2, only the resource name is updated  in the DOC, as mentioned in PR [https://github.com/IBM-Cloud/terraform-provider-ibm/pull/6854](https://github.com/IBM-Cloud/terraform-provider-ibm/pull/6854). The datasource was not modified. Since this documentation refers to a datasource and not a resource, the current content is correct and no changes are required here.

Comment (obai-1) [reply website/docs/d/database.html.markdown:100]: For clarity, both ibm_database resource and datasource will support Gen2.

**Review by obai-1 — APPROVED** [review]
LGTM

**Review by hkantare — APPROVED** [review]
